File: csf_tz/csf_tz/payment_entry.js
frappe.ui.form.on("Payment Entry", {
	onload: function (frm) {
		if (frm.is_new()) {
			frm.trigger("payment_type");
		}
	},
	refresh: function (frm) {
		frm.trigger("add_write_off_button");
	},
	payment_type: function (frm) {
		if (frm.is_new()) {
			if (frm.doc.payment_type == "Receive") {
				frm.set_value("naming_series", "RE-.YYYY.-");
				if (!["Student", "Donor"].includes(frm.doc.party_type)) {
					frm.set_value("party_type", "Customer");
				}
			}
			else if (frm.doc.payment_type == "Pay") {
				frm.set_value("naming_series", "PE-.YYYY.-");
				if (frm.doc.party_type != "Employee") {
					frm.set_value("party_type", "Supplier");
				}
			}
			else if (frm.doc.payment_type == "Internal Transfer") {
				frm.set_value("naming_series", "IT-.YYYY.-");
				frm.set_value("party_type", "");
				frm.set_value("party_name", "");
			}
		}
		frm.refresh_fields()
	},

	party: function (frm) {
		if (frm.is_new()) {
			// check if the feature is disabled in CSF TZ Settings
			frappe.db.get_single_value("CSF TZ Settings", "disable_get_outstanding_functionality")
			.then(disabled => {
				if (disabled) {
					// Feature is disabled, do not proceed with get_outstanding_documents
					return;
				}
				
				// Feature is enabled, proceed with existing functionality
			const today = frappe.datetime.get_today();
			const filters = {
				from_posting_date: frappe.datetime.add_days(today, -3650),
				to_posting_date: today,
				allocate_payment_amount: 1
			}
			if (["Customer", "Supplier"].includes(frm.doc.party_type) && frm.doc.paid_from_account_currency && frm.doc.paid_to_account_currency) {
				frm.events.get_outstanding_documents(frm, filters);
			  }
		    });
		}
	},

	get_outstanding_documents: function (frm, filters) {
		// first check if the feature s disabled in CSF TZ Settings
		return frappe.db.get_single_value("CSF TZ Settings", "disable_get_outstanding_functionality")
			.then(disabled => {
				if (disabled) {
					// Feature is disabled, do not proceed
					return;
				}
				
				// Continue with normal functionality
		if (typeof frappe.route_history[frappe.route_history.length - 2] != "undefined") {
			if (frappe.route_history[frappe.route_history.length - 2][1] in ["Sales Invoice", "Employee Advance", "Purchase Invoice"]) {
				return;
			}
		}

		frm.clear_table("references");

		if (!frm.doc.party) {
			return;
		}

		frm.events.check_mandatory_to_fetch(frm);

		// Ensure party account is set based on payment type
		var party_account = frm.doc.payment_type == "Receive" ? frm.doc.paid_from : frm.doc.paid_to;
		if (!party_account) {
			frappe.msgprint(__("Please set the appropriate account for the selected payment type."));
			return;
		}

		var company_currency = frappe.get_doc(":Company", frm.doc.company).default_currency;

		var args = {
			"posting_date": frm.doc.posting_date,
			"company": frm.doc.company,
			"party_type": frm.doc.party_type,
			"payment_type": frm.doc.payment_type,
			"party": frm.doc.party,
			"party_account": party_account,
			"cost_center": frm.doc.cost_center
		}

		for (let key in filters) {
			args[key] = filters[key];
		}

		frappe.flags.allocate_payment_amount = filters['allocate_payment_amount'];


		return frappe.call({
			method: 'csf_tz.csftz_hooks.payment_entry.get_outstanding_reference_documents',
			args: {
				args: args
			},
			callback: function (r, rt) {
				if (r.message) {
					var total_positive_outstanding = 0;
					var total_negative_outstanding = 0;

					$.each(r.message, function (i, d) {
						var c = frm.add_child("references");
						c.reference_doctype = d.voucher_type;
						c.reference_name = d.voucher_no;
						c.due_date = d.due_date;
						c.posting_date = d.posting_date;
						c.total_amount = d.invoice_amount;
						c.outstanding_amount = d.outstanding_amount;
						c.bill_no = d.bill_no;

						if (!in_list(["Sales Order", "Purchase Order", "Expense Claim", "Fees"], d.voucher_type)) {
							if (flt(d.outstanding_amount) > 0)
								total_positive_outstanding += flt(d.outstanding_amount);
							else
								total_negative_outstanding += Math.abs(flt(d.outstanding_amount));
						}

						var party_account_currency = frm.doc.payment_type == "Receive" ?
							frm.doc.paid_from_account_currency : frm.doc.paid_to_account_currency;

						if (party_account_currency != company_currency) {
							c.exchange_rate = d.exchange_rate;
						} else {
							c.exchange_rate = 1;
						}
						if (in_list(['Sales Invoice', 'Purchase Invoice', "Expense Claim", "Fees"], d.reference_doctype)) {
							c.due_date = d.due_date;
						}
					});

					if (
						(frm.doc.payment_type == "Receive" && frm.doc.party_type == "Customer") ||
						(frm.doc.payment_type == "Pay" && frm.doc.party_type == "Supplier") ||
						(frm.doc.payment_type == "Pay" && frm.doc.party_type == "Employee") ||
						(frm.doc.payment_type == "Receive" && frm.doc.party_type == "Student")
					) {
						if (total_positive_outstanding > total_negative_outstanding)
							if (!frm.doc.paid_amount)
								frm.set_value("paid_amount",
									total_positive_outstanding - total_negative_outstanding);
					} else if (
						total_negative_outstanding &&
						total_positive_outstanding < total_negative_outstanding
					) {
						if (!frm.doc.received_amount)
							frm.set_value("received_amount",
								total_negative_outstanding - total_positive_outstanding);
					}
				}

				const paid_amount = frm.doc.payment_type == "Receive" ? frm.doc.paid_amount : frm.doc.received_amount;
				if (paid_amount) {
					frm.events.allocate_party_amount_against_ref_docs(frm, paid_amount, true);
				}

			}
		  });
		});
	},
	get_outstanding_so: function (frm) {
		const today = frappe.datetime.get_today();
		let fields = [
			{ fieldtype: "Section Break", label: __("Posting Date") },
			{
				fieldtype: "Date",
				label: __("From Date"),
				fieldname: "from_posting_date",
				default: frappe.datetime.add_days(today, -30),
			},
			{ fieldtype: "Column Break" },
			{ fieldtype: "Date", label: __("To Date"), fieldname: "to_posting_date", default: today },
			{ fieldtype: "Section Break", label: __("Due Date") },
			{ fieldtype: "Date", label: __("From Date"), fieldname: "from_due_date" },
			{ fieldtype: "Column Break" },
			{ fieldtype: "Date", label: __("To Date"), fieldname: "to_due_date" },
			{ fieldtype: "Section Break", label: __("Outstanding Amount") },
			{
				fieldtype: "Float",
				label: __("Greater Than Amount"),
				fieldname: "outstanding_amt_greater_than",
				default: 0,
			},
			{ fieldtype: "Column Break" },
			{ fieldtype: "Float", label: __("Less Than Amount"), fieldname: "outstanding_amt_less_than" },
			{ 
				fieldtype: "Check",
				label: __("Allocate Payment Amount"),
				fieldname: "allocate_payment_amount",
				default: 1
			}
		];

		frappe.prompt(
			fields,
			function (filters) {
				frm.clear_table("references");
				
				if (!frm.doc.party) {
					frappe.throw(__("Please select a Party first"));
					return;
				}

				frm.events.check_mandatory_to_fetch(frm);

				// Ensure party account is set based on payment type
				var party_account = frm.doc.payment_type == "Receive" ? frm.doc.paid_from : frm.doc.paid_to;
				if (!party_account) {
					frappe.msgprint(__("Please set the appropriate account for the selected payment type."));
					return;
				}

				frappe.flags.allocate_payment_amount = filters.allocate_payment_amount;
				
				var args = {
					"posting_date": frm.doc.posting_date,
					"company": frm.doc.company,
					"party_type": frm.doc.party_type,
					"payment_type": frm.doc.payment_type,
					"party": frm.doc.party,
					"party_account": party_account,
					"cost_center": frm.doc.cost_center,
					"from_posting_date": filters.from_posting_date,
					"to_posting_date": filters.to_posting_date,
					"from_due_date": filters.from_due_date,
					"to_due_date": filters.to_due_date,
					"outstanding_amt_greater_than": filters.outstanding_amt_greater_than,
					"outstanding_amt_less_than": filters.outstanding_amt_less_than,
					"allocate_payment_amount": filters.allocate_payment_amount
				};

				return frappe.call({
					method: 'csf_tz.csftz_hooks.payment_entry.get_outstanding_sales_orders',
					args: {
						args: args
					},
					callback: function (r, rt) {
						if (r.message) {
							var total_positive_outstanding = 0;

							$.each(r.message, function (i, d) {
								var c = frm.add_child("references");
								c.reference_doctype = d.voucher_type;
								c.reference_name = d.voucher_no;
								c.due_date = d.due_date;
								c.posting_date = d.posting_date;
								c.total_amount = d.invoice_amount;
								c.outstanding_amount = d.outstanding_amount;
								
								// Add to total outstanding
								total_positive_outstanding += flt(d.outstanding_amount);
								
								var party_account_currency = frm.doc.payment_type == "Receive" ?
									frm.doc.paid_from_account_currency : frm.doc.paid_to_account_currency;
								
								var company_currency = frappe.get_doc(":Company", frm.doc.company).default_currency;
								
								if (party_account_currency != company_currency) {
									c.exchange_rate = d.exchange_rate;
								} else {
									c.exchange_rate = 1;
								}
							});

							// Set paid amount based on outstanding sales orders
							if (total_positive_outstanding > 0 && frappe.flags.allocate_payment_amount) {
								if (frm.doc.payment_type == "Receive" && !frm.doc.paid_amount) {
									frm.set_value("paid_amount", total_positive_outstanding);
								} else if (frm.doc.payment_type == "Pay" && !frm.doc.received_amount) {
									frm.set_value("received_amount", total_positive_outstanding);
								}
							}
							
							frm.refresh_fields();
							
							const paid_amount = frm.doc.payment_type == "Receive" ? frm.doc.paid_amount : frm.doc.received_amount;
							if (paid_amount && frappe.flags.allocate_payment_amount) {
								frm.events.allocate_party_amount_against_ref_docs(frm, paid_amount, true);
							}
						}
					}
				});
			},
			__("Filters"),
			__("Get Sales Orders")
		);
	},

	// Write-off Journal Entry Feature
	add_write_off_button: function (frm) {
		// Check if feature is enabled and conditions are met
		frappe.db
			.get_single_value("CSF TZ Settings", "enable_write_off_jv_pe")
			.then((enable_write_off) => {
				if (enable_write_off &&
					frm.doc.docstatus === 1 &&
					frm.doc.unallocated_amount > 0) {

					frm.add_custom_button(__("Write Off Outstanding"), function () {
						// Fetch the write-off account from Company before showing the dialog
						frappe.db.get_value("Company", frm.doc.company, "write_off_account").then(function(r) {
							let write_off_account = r.message ? r.message.write_off_account : null;

							// Show dialog to select write-off account
							let dialog = new frappe.ui.Dialog({
								title: __("Write Off Unallocated Amount"),
								fields: [
									{
										fieldname: "write_off_account",
										label: __("Write Off Account"),
										fieldtype: "Link",
										options: "Account",
										"default": write_off_account,
										reqd: 1,
										get_query: function() {
											return {
												filters: {
													"report_type": "Balance Sheet",
													"is_group": 0,
													"company": frm.doc.company
												}
											};
										}
									},
									{
										fieldname: "unallocated_amount",
										label: __("Unallocated Amount"),
										fieldtype: "Currency",
										default: frm.doc.unallocated_amount,
										read_only: 1
									}
								],
								primary_action_label: __("Create Write Off Entry"),
								primary_action: function(values) {
									frappe.call({
										method: "csf_tz.custom_api.create_write_off_jv_pe",
										args: {
											payment_entry: frm.doc.name,
											account: values.write_off_account
										},
										callback: function(r) {
											if (r.message) {
												const journal_entry_link = `<a href="/app/journal-entry/${encodeURIComponent(r.message)}" target="_blank">${frappe.utils.escape_html(r.message)}</a>`;
												frappe.msgprint(__("Write-off Journal Entry created: {0}", [journal_entry_link]));
												frm.reload_doc();
											}
										}
									});
									dialog.hide();
								}
							});
							dialog.show();
						});
					}, __("Create"));
				}
			});
	}
});

File: csf_tz/csftz_hooks/payment_entry.py
# Copyright (c) 2021, Aakvatech Ltd. and contributors
# For license information, please see license.txt

import json

import frappe
from erpnext.accounts.doctype.payment_entry.payment_entry import (
	get_negative_outstanding_invoices,
	get_orders_to_be_billed,
	split_refdocs_based_on_payment_terms,
)
from erpnext.accounts.utils import get_account_currency, get_outstanding_invoices
from erpnext.controllers.accounts_controller import get_supplier_block_status
from erpnext.setup.utils import get_exchange_rate
from frappe import _, qb
from frappe.utils import getdate, nowdate


@frappe.whitelist()
def get_outstanding_reference_documents(args):
	# Check if the feature is disabled in CSF TZ Settings
	if frappe.db.get_single_value("CSF TZ Settings", "disable_get_outstanding_functionality"):
		return []

	if isinstance(args, str):
		args = json.loads(args)

	if args.get("party_type") == "Member":
		return

	ple = qb.DocType("Payment Ledger Entry")
	common_filter = []
	accounting_dimensions_filter = []
	posting_and_due_date = []

	# confirm that Supplier is not blocked
	if args.get("party_type") == "Supplier":
		supplier_status = get_supplier_block_status(args["party"])
		if supplier_status["on_hold"]:
			if supplier_status["hold_type"] == "All":
				return []
			elif supplier_status["hold_type"] == "Payments":
				if (
					not supplier_status["release_date"]
					or getdate(nowdate()) <= supplier_status["release_date"]
				):
					return []

	party_account_currency = get_account_currency(args.get("party_account"))
	company_currency = frappe.get_cached_value("Company", args.get("company"), "default_currency")

	# Get positive outstanding sales /purchase invoices
	condition = ""
	if args.get("voucher_type") and args.get("voucher_no"):
		condition = " and voucher_type={} and voucher_no={}".format(
			frappe.db.escape(args["voucher_type"]), frappe.db.escape(args["voucher_no"])
		)
		common_filter.append(ple.voucher_type == args["voucher_type"])
		common_filter.append(ple.voucher_no == args["voucher_no"])

	# Add cost center condition
	# if args.get("cost_center"):
	# 	condition += " and cost_center='%s'" % args.get("cost_center")
	# 	accounting_dimensions_filter.append(ple.cost_center == args.get("cost_center"))

	date_fields_dict = {
		"posting_date": ["from_posting_date", "to_posting_date"],
		"due_date": ["from_due_date", "to_due_date"],
	}

	for fieldname, date_fields in date_fields_dict.items():
		if args.get(date_fields[0]) and args.get(date_fields[1]):
			condition += (
				f" and {fieldname} between '{args.get(date_fields[0])}' and '{args.get(date_fields[1])}'"
			)
			posting_and_due_date.append(ple[fieldname][args.get(date_fields[0]) : args.get(date_fields[1])])

	if args.get("company"):
		condition += " and company = {}".format(frappe.db.escape(args.get("company")))
		common_filter.append(ple.company == args.get("company"))

	party_account = [args.get("party_account")]
	outstanding_invoices = get_outstanding_invoices(
		args.get("party_type"),
		args.get("party"),
		party_account,
		common_filter=common_filter,
		posting_date=posting_and_due_date,
		min_outstanding=args.get("outstanding_amt_greater_than"),
		max_outstanding=args.get("outstanding_amt_less_than"),
		accounting_dimensions=accounting_dimensions_filter,
	)
	outstanding_invoices = split_refdocs_based_on_payment_terms(outstanding_invoices, args.get("company"))

	for d in outstanding_invoices:
		d["exchange_rate"] = 1
		d["posting_date"] = frappe.db.get_value(d.voucher_type, d.voucher_no, "posting_date")
		if party_account_currency != company_currency:
			if d.voucher_type in frappe.get_hooks("invoice_doctypes"):
				d["exchange_rate"] = frappe.db.get_value(d.voucher_type, d.voucher_no, "conversion_rate")
			elif d.voucher_type == "Journal Entry":
				d["exchange_rate"] = get_exchange_rate(
					party_account_currency, company_currency, d.posting_date
				)
		if d.voucher_type in ("Purchase Invoice"):
			d["bill_no"] = frappe.db.get_value(d.voucher_type, d.voucher_no, "bill_no")

	# Get all SO / PO which are not fully billed or against which full advance not paid
	orders_to_be_billed = []
	# orders_to_be_billed = get_orders_to_be_billed(
	#     args.get("posting_date"),
	#     args.get("party_type"),
	#     args.get("party"),
	#     args.get("company"),
	#     party_account_currency,
	#     company_currency,
	#     filters=args,
	# )

	# Get negative outstanding sales /purchase invoices
	negative_outstanding_invoices = []
	if args.get("party_type") != "Employee" and not args.get("voucher_no"):
		negative_outstanding_invoices = get_negative_outstanding_invoices(
			args.get("party_type"),
			args.get("party"),
			args.get("party_account"),
			party_account_currency,
			company_currency,
			condition=condition,
		)

	data = negative_outstanding_invoices + outstanding_invoices + orders_to_be_billed

	if not data:
		frappe.msgprint(
			_(
				"No outstanding invoices found for the {0} {1} which qualify the filters you have specified."
			).format(_(args.get("party_type")).lower(), frappe.bold(args.get("party")))
		)

	return data


@frappe.whitelist()
def get_outstanding_sales_orders(args):
	if isinstance(args, str):
		args = json.loads(args)

	if args.get("party_type") == "Member":
		return

	if args.get("party_type") != "Customer":
		frappe.throw(_("Sales Orders can only be fetched for Customer"))

	# confirm that Supplier is not blocked (not needed for Customer, but keeping for consistency)
	if args.get("party_type") == "Supplier":
		supplier_status = get_supplier_block_status(args["party"])
		if supplier_status["on_hold"]:
			if supplier_status["hold_type"] == "All":
				return []
			elif supplier_status["hold_type"] == "Payments":
				if (
					not supplier_status["release_date"]
					or getdate(nowdate()) <= supplier_status["release_date"]
				):
					return []

	party_account_currency = get_account_currency(args.get("party_account"))
	company_currency = frappe.get_cached_value("Company", args.get("company"), "default_currency")

	# Get all SO which are not fully billed or against which full advance not paid
	orders_to_be_billed = get_orders_to_be_billed(
		args.get("posting_date"),
		args.get("party_type"),
		args.get("party"),
		args.get("company"),
		party_account_currency,
		company_currency,
		filters=args,
	)

	for d in orders_to_be_billed:
		d["exchange_rate"] = 1
		if party_account_currency != company_currency:
			if d.voucher_type in frappe.get_hooks("invoice_doctypes"):
				d["exchange_rate"] = frappe.db.get_value(d.voucher_type, d.voucher_no, "conversion_rate")
			else:
				d["exchange_rate"] = get_exchange_rate(
					party_account_currency, company_currency, d.posting_date
				)

	# Add posting date
	for d in orders_to_be_billed:
		d["posting_date"] = frappe.db.get_value(
			d.voucher_type,
			d.voucher_no,
			"transaction_date" if d.voucher_type == "Sales Order" else "posting_date",
		)

		# Add due date (if available)
		if d.voucher_type == "Sales Order":
			d["due_date"] = frappe.db.get_value(d.voucher_type, d.voucher_no, "delivery_date") or ""

	if not orders_to_be_billed:
		frappe.msgprint(
			_(
				"No outstanding Sales Orders found for the {0} {1} which qualify the filters you have specified."
			).format(_(args.get("party_type")).lower(), frappe.bold(args.get("party")))
		)

	return orders_to_be_billed


def validate(self, method):
	company = frappe.get_cached_doc("Company", self.company)

	if company.restrict_unallocated_amount_for_supplier and self.restrict_unallocated_amount_for_supplier:
		if self.unallocated_amount > 0:
			frappe.throw(
				_("Cannot submit Payment Entry <b>{0}</b> with unallocated amount.").format(self.name)
			)
